pkg/store/gcworker/gc_worker.go

```diff
@@ -805,73 +806,93 @@ func (w *GCWorker) runGCJob(ctx context.Context, safePoint uint64, concurrency i
+				task := rangetask.NewDeleteRangeTask(w.tikvStore, startKey, endKey, concurrency)
+				err = task.Execute(ctx)
+			} else {
+				err = w.doUnsafeDestroyRangeRequest(ctx, startKey, endKey, concurrency)
+			}
+			failpoint.Inject("ignoreDeleteRangeFailed", func() {
+				err = nil
+			})
 
-		if err != nil {
-			logutil.Logger(ctx).Error("delete range failed on range", zap.String("category", "gc worker"),
-				zap.String("uuid", w.uuid),
-				zap.Stringer("startKey", startKey),
-				zap.Stringer("endKey", endKey),
-				zap.Error(err))
-			continue
-		}
+			if err != nil {
+				logutil.Logger(ctx).Error("delete range failed on range", zap.String("category", "gc worker"),
+					zap.String("uuid", w.uuid),
+					zap.Stringer("startKey", startKey),
+					zap.Stringer("endKey", endKey),
+					zap.Error(err))
+				return
+			}
 
-		if err := w.doGCPlacementRules(se, safePoint, r, gcPlacementRuleCache); err != nil {
-			logutil.Logger(ctx).Error("gc placement rules failed on range", zap.String("category", "gc worker"),
-				zap.String("uuid", w.uuid),
-				zap.Int64("jobID", r.JobID),
-				zap.Int64("elementID", r.ElementID),
-				zap.Error(err))
-			continue
-		}
-		if err := w.doGCLabelRules(r); err != nil {
-			logutil.Logger(ctx).Error("gc label rules failed on range", zap.String("category", "gc worker"),
-				zap.String("uuid", w.uuid),
-				zap.Int64("jobID", r.JobID),
-				zap.Int64("elementID", r.ElementID),
-				zap.Error(err))
-			continue
-		}
+			err = doGCPlacementRules(se, safePoint, r, gcPlacementRuleCache)
+			if err != nil {
+				logutil.Logger(ctx).Error("gc placement rules failed on range", zap.String("category", "gc worker"),
+					zap.String("uuid", w.uuid),
+					zap.Int64("jobID", r.JobID),
+					zap.Int64("elementID", r.ElementID),
+					zap.Error(err))
+			}
+			if err := w.doGCLabelRules(r); err != nil {
+				logutil.Logger(ctx).Error("gc label rules failed on range", zap.String("category", "gc worker"),
+					zap.String("uuid", w.uuid),
+					zap.Int64("jobID", r.JobID),
+					zap.Int64("elementID", r.ElementID),
+					zap.Error(err))
+				return
+			}
 
-		err = util.CompleteDeleteRange(se, r, !v2)
-		if err != nil {
-			logutil.Logger(ctx).Error("failed to mark delete range task done", zap.String("category", "gc worker"),
-				zap.String("uuid", w.uuid),
-				zap.Stringer("startKey", startKey),
-				zap.Stringer("endKey", endKey),
-				zap.Error(err))
-			metrics.GCUnsafeDestroyRangeFailuresCounterVec.WithLabelValues("save").Inc()
+			err = util.CompleteDeleteRange(se, r, !v2)
+			if err != nil {
+				logutil.Logger(ctx).Error("failed to mark delete range task done", zap.String("category", "gc worker"),
+					zap.String("uuid", w.uuid),
+					zap.Stringer("startKey", startKey),
+					zap.Stringer("endKey", endKey),
+					zap.Error(err))
+				metrics.GCUnsafeDestroyRangeFailuresCounterVec.WithLabelValues("save").Inc()
+			}
 		}
+		wg.Run(func() { f(r) })
```

Comment:
Is there an issue of [loop var](https://go.dev/blog/loopvar-preview)?

Reply:
Good catch. I paid attention to this but it was broken when refactoring from `go func(){}(r)` to `wg.Run(func(){f(r)})`...